# Adult Census Income Dataset — Data Exploration

## Machine Learning Fundamentals

### Objective

The goal of this notebook is to understand the Adult Census Income dataset through Exploratory Data Analysis (EDA).

We will analyze:

- Dataset structure
- Features and target variable
- Data types
- Missing values
- Statistical properties
- Feature distributions
- Relationships between variables

The target variable is:
Income

In [15]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

import sys
from pathlib import Path

sys.path.append(str(Path.cwd().parent))

from src.data_loader import load_dataset

## 1. Dataset Loading

The Adult Census Income dataset contains demographic and employment information collected from census data.

The dataset contains:

- Numerical features
- Categorical features
- Target income class

The dataset does not contain predefined column names, therefore column names are assigned during loading.

In [16]:
dataset_path = "../datasets/adult/adult.data"

df = load_dataset(dataset_path)

df.head()

,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K


## 2. Dataset Overview

The first step in machine learning is understanding the dataset structure.

We will examine:

- Number of rows
- Number of columns
- Feature names
- Sample records

In [17]:
df.shape

(32561, 15)

In [18]:

df.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education_num',
       'marital_status', 'occupation', 'relationship', 'race', 'sex',
       'capital_gain', 'capital_loss', 'hours_per_week', 'native_country',
       'income'],
      dtype='str')

In [19]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   age             32561 non-null  int64
 1   workclass       30725 non-null  str  
 2   fnlwgt          32561 non-null  int64
 3   education       32561 non-null  str  
 4   education_num   32561 non-null  int64
 5   marital_status  32561 non-null  str  
 6   occupation      30718 non-null  str  
 7   relationship    32561 non-null  str  
 8   race            32561 non-null  str  
 9   sex             32561 non-null  str  
 10  capital_gain    32561 non-null  int64
 11  capital_loss    32561 non-null  int64
 12  hours_per_week  32561 non-null  int64
 13  native_country  31978 non-null  str  
 14  income          32561 non-null  str  
dtypes: int64(6), str(9)
memory usage: 3.7 MB


## Dataset Information Findings

- The Adult Census dataset contains 32,561 records and 15 features.
- The dataset contains both numerical and categorical variables.
- Numerical variables do not contain missing values.
- Three categorical features contain missing values:
    - workclass
    - occupation
    - native_country
- Missing values were originally represented as "?" and converted into NaN during data loading.
- Further missing value handling will be performed in the data preprocessing stage.

In [20]:
df.describe()

,age,fnlwgt,education_num,capital_gain,capital_loss,hours_per_week
count,32561.000000,3.256100e+04,32561.000000,32561.000000,32561.000000,32561.000000
mean,38.581647,1.897784e+05,10.080679,1077.648844,87.303830,40.437456
std,13.640433,1.055500e+05,2.572720,7385.292085,402.960219,12.347429
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178270e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.370510e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


## Numerical Feature Statistical Analysis

The `describe()` function provides statistical information about numerical features.

It includes:

- Count
- Mean
- Standard deviation
- Minimum value
- Quartiles (25%, 50%, 75%)
- Maximum value

These statistics help us understand:

- Data distribution
- Feature ranges
- Possible outliers
- Central tendency

In [21]:
df.describe(include="object")

C:\Users\abuba\AppData\Local\Temp\ipykernel_11848\702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,workclass,education,marital_status,occupation,relationship,race,sex,native_country,income
count,30725,32561,32561,30718,32561,32561,32561,31978,32561
unique,8,16,7,14,6,5,2,41,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22696,10501,14976,4140,13193,27816,21790,29170,24720


## Categorical Feature Statistical Analysis

Categorical features contain non-numerical information such as:
## Categorical Feature Findings Summary

From categorical feature analysis:

- The dataset contains multiple demographic and employment-related categorical variables.
- Private sector workers represent the majority of workclass records.
- HS-grad is the most common education level.
- United-States is the dominant native country.
- Missing values are present in:
    - workclass
    - occupation
    - native_country

The target variable `income` contains two classes:

- <=50K
- >50K

The target distribution appears imbalanced, which should be considered during model development.
The `describe(include="object")` function provides:

- Count of available values
- Number of unique categories
- Most frequent category
- Frequency of the most common category

## Unique Value Analysis

The number of unique values in each feature helps us understand the complexity of the dataset.

Low-cardinality features usually represent categories and can be encoded using techniques such as:

- One-Hot Encoding
- Label Encoding

High-cardinality features may require special consideration during preprocessing.

In [22]:
df.nunique()

age                  73
workclass             8
fnlwgt            21648
education            16
education_num        16
marital_status        7
occupation           14
relationship          6
race                  5
sex                   2
capital_gain        119
capital_loss         92
hours_per_week       94
native_country       41
income                2
dtype: int64

## Unique Value Findings Summary

- Numerical features contain different value ranges and distributions.
- `fnlwgt` has a high number of unique values because it represents census weights.
- `education` contains ordered categories and may benefit from ordinal encoding.
- Most categorical features have low cardinality and can be handled using one-hot encoding.
- The target variable `income` contains two classes, making this a binary classification problem.

Feature cardinality analysis will guide the preprocessing strategy in later stages.

In [23]:
df.duplicated().sum()

np.int64(24)

## Duplicate Records Analysis

Duplicate records can negatively affect machine learning models by:

- Increasing bias in the dataset
- Giving repeated importance to certain samples
- Affecting statistical analysis

We check duplicate rows to understand dataset quality.

In [26]:
df.sample(10)


,age,workclass,fnlwgt,education,education_num,marital_status,occupation,relationship,race,sex,capital_gain,capital_loss,hours_per_week,native_country,income
22455,40,Private,48087,HS-grad,9,Married-civ-spouse,Sales,Husband,White,Male,7298,0,45,United-States,>50K
2887,19,Federal-gov,30559,HS-grad,9,Married-AF-spouse,Sales,Wife,White,Female,0,0,40,United-States,<=50K
16338,38,Self-emp-not-inc,37778,HS-grad,9,Never-married,Farming-fishing,Not-in-family,White,Male,0,0,60,United-States,<=50K
6292,18,Private,302859,HS-grad,9,Never-married,Sales,Own-child,White,Male,0,0,20,United-States,<=50K
31596,39,Private,72338,Masters,14,Married-civ-spouse,Prof-specialty,Husband,Asian-Pac-Islander,Male,0,0,40,Philippines,>50K
25505,28,Private,168514,Assoc-acdm,12,Married-civ-spouse,Exec-managerial,Wife,White,Female,0,0,40,United-States,<=50K
20682,68,Private,154897,HS-grad,9,Never-married,Sales,Not-in-family,White,Female,0,0,30,United-States,<=50K
7080,18,Private,74054,11th,7,Never-married,Sales,Own-child,Other,Female,0,0,20,NaN,<=50K
14948,39,Self-emp-not-inc,336793,Masters,14,Married-civ-spouse,Craft-repair,Husband,White,Male,0,0,55,United-States,>50K
12534,24,State-gov,232918,Bachelors,13,Never-married,Prof-specialty,Not-in-family,White,Male,0,0,20,United-States,<=50K
